plugin.yaml

```diff
@@ -10,62 +10,281 @@ plugins:
+              default: { get_property: [SELF, use_external_resource] }
+        delete:
+          implementation: gcp_plugin.plugin.tasks.delete_keypair
+          inputs:
+            gcp_config: 
+              default: { get_property: [SELF, gcp_config] }
+            user: 
+              default: { get_property: [SELF, user] }
+            private_key_path: 
+              default: { get_property: [SELF, private_key_path] }
+
+  cloudify.gcp.nodes.ExternalIP:
+    derived_from: cloudify.nodes.VirtualIP
+    properties:
+      use_external_resource:
+        description: >
+          Indicate whether the resource exists or if Cloudify should create the resource.
+        type: boolean
+        default: false
+      gcp_config:
+        description: >
+          A dictionary of values to pass to authenticate with the GCP API.
+        default: {}
+
 relationships:
-  cloudify.gcp.relationships.connected_to:
+  cloudify.gcp.relationships.instance_connected_to_security_group:
     derived_from: cloudify.relationships.connected_to
+    source_interfaces:
+      cloudify.interfaces.relationship_lifecycle:
+        establish:
+          implementation: gcp_plugin.plugin.tasks.add_instance_tag
+          inputs:
+            gcp_config:
+              default: { get_property: [SOURCE, gcp_config] }
+            instance_name: 
+              default: { get_attribute: [SOURCE, gcp_name] }
+            tag:
+              default: { get_attribute: [TARGET, targetTags] }
+        unlink:
+          implementation: gcp_plugin.plugin.tasks.remove_instance_tag
+          inputs:
+            gcp_config:
+              default: { get_property: [SOURCE, gcp_config] }
+            instance_name: 
+              default: { get_attribute: [SOURCE, gcp_name] }
+            tag:
+              default: { get_attribute: [TARGET, targetTags] }
 
-  cloudify.gcp.relationships.contained_in:
-    derived_from: cloudify.relationships.contained_in
+  cloudify.gcp.relationships.instance_connected_to_ip:
+    derived_from: cloudify.relationships.connected_to
+    source_interfaces:
+      cloudify.interfaces.relationship_lifecycle:
+        establish:
+          implementation: gcp_plugin.plugin.tasks.add_external_ip
+          inputs:
+            gcp_config:
+              default: { get_property: [SOURCE, gcp_config] }
+            instance_name: 
+              default: { get_attribute: [SOURCE, gcp_name] }
+        unlink:
+          implementation: gcp_plugin.plugin.tasks.remove_external_ip
+          inputs:
+            gcp_config:
+              default: { get_property: [SOURCE, gcp_config] }
+            instance_name: 
+              default: { get_attribute: [SOURCE, gcp_name] }
 
+  cloudify.gcp.relationships.instance_connected_to_keypair:
+    derived_from: cloudify.relationships.connected_to
+    source_interfaces:
+      cloudify.interfaces.relationship_lifecycle:
+        establish:
+          implementation: gcp_plugin.plugin.tasks.save_private_key
```

Comment:
I think you don't need to explicity define the establish step and make it call save_private_key, because the key pair is created and the private key is saved on the create_keypair operation. This probably means you can just declare in the manager blueprint that there's a relationship between the server node and the keypair node, and they will be created in the correct order.